### DISCUSSION TITLE
About Dotnet IoT use this to convert images to RGB565

### DISCUSSION THREAD

**Discussion by GreenShadeZhang** [SixLabors/ImageSharp#2296] [Q&A]
At present, I use SPI communication to implement an LCD screen driver, the screen data is RGB565, when I use this library for image processing, I found that only BGR565, I use BGR565 to transmit, the picture can be displayed normally, but the picture color is not normal, I also tried to use traversed pixels to operate but it seems to be written incorrectly, I want to ask how this situation should deal with pixel data.

**Comment by brianpopow** [comment]
If the LCD screen expects RGB565 you cant send BGR565 and expect it to work.

ImageSharp does not have this pixel format, but you could try to use an implementation from the [ImageSharp.Texture](https://github.com/SixLabors/ImageSharp.Textures) project: https://github.com/SixLabors/ImageSharp.Textures/blob/main/src/ImageSharp.Textures/PixelFormats/Generated/Rgb565.cs

Just copy the file to your project and see if that works.

**Reply by GreenShadeZhang** [reply]
> If the LCD screen expects RGB565 you cant send BGR565 and expect it to work.
> 
> ImageSharp does not have this pixel format, but you could try to use an implementation from the [ImageSharp.Texture](https://github.com/SixLabors/ImageSharp.Textures) project: https://github.com/SixLabors/ImageSharp.Textures/blob/main/src/ImageSharp.Textures/PixelFormats/Generated/Rgb565.cs
> 
> Just copy the file to your project and see if that works.


When I used the test code for testing, the effect was still not ideal, there was noise, at first I thought that I had a problem with the way I drove the driver, but the image was normal when I used python code to put arrays everywhere and write them in C#.

My code address
https://github.com/GreenShadeZhang/dotnet-iot-tutorial-code/tree/main/src/ST7789V3/ST7789V3

![QQ图片20221114222955](https://user-images.githubusercontent.com/24240675/201686390-36906c3b-546f-4a79-846b-ad32396df57e.jpg)
![QQ图片20221114223004](https://user-images.githubusercontent.com/24240675/201686405-0f7b4041-9ead-438c-ae21-a1f3fa661f51.jpg)

Reference python code
```python
        
    def ShowImage(self,Image):
        """Set buffer to value of Python Imaging Library image."""
        """Write display buffer to physical display"""
                
        imwidth, imheight = Image.size
        if imwidth != self.width or imheight != self.height:
            raise ValueError('Image must be same dimensions as display \
                ({0}x{1}).' .format(self.width, self.height))
        img = self.np.asarray(Image)
        pix = self.np.zeros((self.height,self.width,2), dtype = self.np.uint8)
        
        pix[...,[0]] = self.np.add(self.np.bitwise_and(img[...,[0]],0xF8),self.np.right_shift(img[...,[1]],5))
        pix[...,[1]] = self.np.add(self.np.bitwise_and(self.np.left_shift(img[...,[1]],3),0xE0),self.np.right_shift(img[...,[2]],3))
        
        pix = pix.flatten().tolist()
        #print(pix)
        print(len(pix))
        self.SetWindows ( 0, 0, self.width, self.height)
        self.digital_write(self.DC_PIN,self.GPIO.HIGH)
        for i in range(0,len(pix),4096):
            self.spi_writebyte(pix[i:i+4096])		
```

**Reply by brianpopow** [reply]
@GreenShadeZhang I do not understand what you mean. Have you tried to use the Rgb565 pixel class from the textures project and with that you get those artifacts?

If there is an issue with it, I would assume to look into the `ToVector4` and `Pack` method first. Maybe there is an issue with it, but I can not spot something obvious in it.

**Reply by GreenShadeZhang** [reply]
> @GreenShadeZhang I do not understand what you mean. Have you tried to use the Rgb565 pixel class from the textures project and with that you get those artifacts?
> 
> If there is an issue with it, I would assume to look into the `ToVector4` and `Pack` method first. Maybe there is an issue with it, but I can not spot something obvious in it.

I mean I initially suspected that there was something wrong with my driver writing, so I used the reference python code to export the byte array directly to my driver to display it was normal.

When I use ImageSharp, the image will show noise, I followed your advice to use RGB565 is the same result, the python code I just posted is not very clear about the role of that bit manipulation, so I was wondering if the python code I referenced did a separate operation on pixels.

**Reply by brianpopow** [reply]
Can you post the original image here? What format is it? Maybe there is something wrong decoding the image.

**Reply by GreenShadeZhang** [reply]
> Can you post the original image here? What format is it? Maybe there is something wrong decoding the image.

I don't think it's a decoding issue, after I use bgr565 to read the image to byte[] and then save to the file, I see the picture is normal.
[LCD_1inch47.zip](https://github.com/SixLabors/ImageSharp/files/10024200/LCD_1inch47.zip)

**Reply by JimBobSquarePants** [reply]
Normal?

**Reply by GreenShadeZhang** [reply]
> Normal?

Yes, it's normal, so I want to understand what the Python code does, why the data after Python transcoding is normal, and I plan to have time to use OpenCVsharp to parse the image into byte[] and write it for verification.

**Reply by JimBobSquarePants** [reply]
Ah... I understand. When you save to a file the pixel format will be converted to the format used by the codec. 

The pixel format represents the layout of the expanded image which is unrelated to the encoded format.

This all reads to me like there's a bug in the Textures version of Rgb565. I would suggest copying our Bgr565 implementation and switching the components. We should probably just add it to the default pixel format collection.

**Reply by GreenShadeZhang** [reply]
> Ah... I understand. When you save to a file the pixel format will be converted to the format used by the codec.
> 
> The pixel format represents the layout of the expanded image which is unrelated to the encoded format.
> 
> This all reads to me like there's a bug in the Textures version of Rgb565. I would suggest copying our Bgr565 implementation and switching the components. We should probably just add it to the default pixel format collection.

Do you mean let me refer to BGR565 to implement an own RGB565?

**Reply by JimBobSquarePants** [reply]
Yep. That’s probably wise
